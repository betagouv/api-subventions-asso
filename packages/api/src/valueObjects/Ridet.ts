import Rid from "./Rid";

export default class Ridet {
    private ridet: string;

    constructor(ridet: string) {
        if (!Ridet.isRidet(ridet)) {
            throw new Error("Invalid Ridet: " + ridet);
        }
        this.ridet = ridet;
    }

    static isRidet(ridet: string | null | undefined): boolean {
        return typeof ridet === "string" && /^\d{9,10}$/.test(ridet);
    }

    get name(): "ridet" {
        return "ridet";
    }

    get value() {
        return this.ridet;
    }

    toRid() {
        return new Rid(this.ridet.slice(0, -3));
    }
}
