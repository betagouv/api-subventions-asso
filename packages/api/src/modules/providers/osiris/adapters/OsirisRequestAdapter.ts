import {
    CommonApplicationDto,
    ApplicationStatus,
    Association,
    DemandeSubvention,
    Etablissement,
    ProviderValue,
    RnaDto,
} from "dto";
import { siretToNIC, siretToSiren } from "../../../../shared/helpers/SirenHelper";
import ProviderValueFactory from "../../../../shared/ProviderValueFactory";
import OsirisActionEntity from "../entities/OsirisActionEntity";
import OsirisRequestEntity from "../entities/OsirisRequestEntity";
import osirisService from "../osiris.service";
import { toStatusFactory } from "../../providers.adapter";
import { RawApplication } from "../../../grant/@types/rawGrant";
import { ApplicationFlatEntity, ApplicationNature } from "../../../../entities/ApplicationFlatEntity";
import Siret from "../../../../identifierObjects/Siret";
import Ridet from "../../../../identifierObjects/Ridet";
import { GenericParser } from "../../../../shared/GenericParser";

export default class OsirisRequestAdapter {
    static PROVIDER_NAME = "Osiris";
    private static _statusConversionArray: { label: ApplicationStatus; providerStatusList: string[] }[] = [
        { label: ApplicationStatus.REFUSED, providerStatusList: ["Refusé"] },
        {
            label: ApplicationStatus.GRANTED,
            providerStatusList: ["Traitement Sirepa", "Traitement Chorus", "Terminé", "A évaluer"],
        },
        { label: ApplicationStatus.INELIGIBLE, providerStatusList: ["Rejeté", "Supprimé"] },
        {
            label: ApplicationStatus.PENDING,
            providerStatusList: [
                "Edition document",
                "Renvoyé au compte asso",
                "En cours d'instruction",
                "En attente superviseur",
                "En attente décision",
            ],
        },
    ];

    private static readonly toStatus = toStatusFactory(OsirisRequestAdapter._statusConversionArray);

    static toAssociation(entity: OsirisRequestEntity, actions: OsirisActionEntity[] = []): Association {
        const dataDate = new Date(Date.UTC(entity.providerInformations.exercise, 0));
        const toPVs = ProviderValueFactory.buildProviderValuesAdapter(OsirisRequestAdapter.PROVIDER_NAME, dataDate);
        const federation =
            actions.length &&
            actions.find(action => action.indexedInformations.federation)?.indexedInformations.federation;
        const licencies =
            actions.length &&
            actions.find(action => action.indexedInformations.federation)?.indexedInformations.licencies;
        const licenciesHommes =
            actions.length &&
            actions.find(action => action.indexedInformations.federation)?.indexedInformations.licenciesHommes;
        const licenciesFemmes =
            actions.length &&
            actions.find(action => action.indexedInformations.federation)?.indexedInformations.licenciesFemmes;

        return {
            siren: toPVs(siretToSiren(entity.legalInformations.siret)),
            rna: entity.legalInformations.rna == undefined ? undefined : toPVs(entity.legalInformations.rna as RnaDto),
            denomination_rna: toPVs(entity.legalInformations.name),
            etablisements_siret: toPVs([entity.legalInformations.siret]),
            nic_siege: entity.providerInformations.etablissementSiege
                ? toPVs(siretToNIC(entity.legalInformations.siret))
                : undefined,
            federation: federation ? toPVs(federation) : undefined,
            licencies:
                licencies && licenciesHommes && licenciesFemmes
                    ? {
                          total: toPVs(licencies),
                          hommes: toPVs(licenciesHommes),
                          femmes: toPVs(licenciesFemmes),
                      }
                    : undefined,
            ...(actions.length
                ? {
                      benevoles: {
                          nombre: toPVs(actions[0].indexedInformations.benevoles),
                          ETPT: toPVs(actions[0].indexedInformations.benevolesETPT),
                      },
                      salaries: {
                          nombre: toPVs(actions[0].indexedInformations.salaries),
                          cdi: toPVs(actions[0].indexedInformations.salariesCDI),
                          cdiETPT: toPVs(actions[0].indexedInformations.salariesCDIETPT),
                          cdd: toPVs(actions[0].indexedInformations.salariesCDD),
                          cddETPT: toPVs(actions[0].indexedInformations.salariesCDDETPT),
                          emploisAides: toPVs(actions[0].indexedInformations.emploiesAides),
                          emploisAidesETPT: toPVs(actions[0].indexedInformations.emploiesAidesETPT),
                      },
                      volontaires: {
                          nombre: toPVs(actions[0].indexedInformations.volontaires),
                          ETPT: toPVs(actions[0].indexedInformations.volontairesETPT),
                      },
                  }
                : {}),
        };
    }

    static toEtablissement(entity: OsirisRequestEntity): Etablissement {
        const dataDate = new Date(Date.UTC(entity.providerInformations.exercise, 0));
        const toPVs = ProviderValueFactory.buildProviderValuesAdapter(OsirisRequestAdapter.PROVIDER_NAME, dataDate);

        return {
            siret: toPVs(entity.legalInformations.siret),
            nic: toPVs(siretToNIC(entity.legalInformations.siret)),
            siege: toPVs(entity.providerInformations.etablissementSiege),
            adresse: toPVs({
                voie: entity.providerInformations.etablissementVoie,
                code_postal: entity.providerInformations.etablissementCodePostal,
                commune: entity.providerInformations.etablissementCommune,
            }),
            representants_legaux: [
                toPVs({
                    nom: entity.providerInformations.representantNom,
                    prenom: entity.providerInformations.representantPrenom,
                    civilite: entity.providerInformations.representantCivilite,
                    role: entity.providerInformations.representantRole,
                    telephone: entity.providerInformations.representantPhone,
                    email: entity.providerInformations.representantEmail,
                }),
            ],
            contacts: [
                toPVs({
                    nom: entity.providerInformations.representantNom,
                    prenom: entity.providerInformations.representantPrenom,
                    civilite: entity.providerInformations.representantCivilite,
                    role: entity.providerInformations.representantRole,
                    telephone: entity.providerInformations.representantPhone,
                    email: entity.providerInformations.representantEmail,
                }),
            ],
            information_banquaire:
                entity.providerInformations.etablissementBIC && entity.providerInformations.etablissementIBAN
                    ? [
                          toPVs({
                              bic: entity.providerInformations.etablissementBIC,
                              iban: entity.providerInformations.etablissementIBAN,
                          }),
                      ]
                    : [],
        };
    }

    static rawToApplication(rawApplication: RawApplication<OsirisRequestEntity>) {
        return this.toDemandeSubvention(rawApplication.data);
    }

    static toDemandeSubvention(entity: OsirisRequestEntity): DemandeSubvention {
        const dataDate = new Date(Date.UTC(entity.providerInformations.exercise, 0));
        const toPV = ProviderValueFactory.buildProviderValueAdapter(osirisService.provider.name, dataDate);

        const EJ = entity.providerInformations.ej ? toPV(entity.providerInformations.ej) : undefined;

        const data: DemandeSubvention = {
            annee_demande: toPV(entity.providerInformations.exercise),
            siret: toPV(entity.legalInformations.siret),
            service_instructeur: toPV(entity.providerInformations.service_instructeur),
            dispositif: toPV(entity.providerInformations.dispositif),
            sous_dispositif: toPV(entity.providerInformations.sous_dispositif),
            statut_label: toPV(OsirisRequestAdapter.toStatus(entity.providerInformations.status)),
            status: toPV(entity.providerInformations.status),
            pluriannualite: toPV(entity.providerInformations.pluriannualite),
            ej: EJ,
            versementKey: EJ,
            date_commision: entity.providerInformations.dateCommission
                ? toPV(entity.providerInformations.dateCommission)
                : undefined,
            contact: {
                email: toPV(entity.providerInformations.representantEmail),
                telephone: entity.providerInformations.representantPhone
                    ? toPV(entity.providerInformations.representantPhone)
                    : undefined,
            },
            montants: {
                total: toPV(entity.providerInformations.montantsTotal),
                demande: toPV(entity.providerInformations.montantsDemande),
                propose: toPV(entity.providerInformations.montantsPropose),
                accorde: toPV(entity.providerInformations.montantsAccorde),
            },
            versement: {
                acompte: toPV(entity.providerInformations.versementAcompte),
                solde: toPV(entity.providerInformations.versementSolde),
                realise: toPV(entity.providerInformations.versementRealise),
                compensation: {
                    "n-1": toPV(entity.providerInformations.versementCompensationN1),
                    reversement: toPV(entity.providerInformations.versementCompensationN),
                },
            },
        };

        if (entity.actions) {
            const territoires = entity.actions.map(action => {
                return {
                    status: toPV(action.indexedInformations.territoireStatus),
                    commentaire: toPV(action.indexedInformations.territoireCommentaire),
                };
            });

            data.territoires = territoires.reduce(
                (acc, territoire) => {
                    if (
                        acc.some(
                            t =>
                                t.status.value === territoire.status.value &&
                                t.status.last_update === territoire.status.last_update &&
                                t.commentaire.value === territoire.commentaire.value &&
                                t.commentaire.last_update === territoire.commentaire.last_update,
                        )
                    )
                        return acc;

                    return acc.concat(territoires);
                },
                [] as { status: ProviderValue<string>; commentaire: ProviderValue<string> }[],
            );

            data.actions_proposee = entity.actions.map(action => ({
                ej: action.indexedInformations.ej ? toPV(action.indexedInformations.ej) : undefined,
                rang: toPV(action.indexedInformations.rang),
                intitule: toPV(action.indexedInformations.intitule),
                objectifs: toPV(action.indexedInformations.objectifs),
                objectifs_operationnels: toPV(action.indexedInformations.objectifs_operationnels),
                description: toPV(action.indexedInformations.description),
                nature_aide: toPV(action.indexedInformations.nature_aide),
                modalite_aide: toPV(action.indexedInformations.modalite_aide),
                modalite_ou_dispositif: toPV(action.indexedInformations.modalite_ou_dispositif),
                indicateurs: toPV(action.indexedInformations.indicateurs),
                cofinanceurs: {
                    noms: toPV(action.indexedInformations.cofinanceurs),
                    montant_demandes: toPV(action.indexedInformations.cofinanceurs_montant_demandes),
                },
                montants_versement: {
                    total: toPV(action.indexedInformations.montants_versement_total),
                    demande: toPV(action.indexedInformations.montants_versement_demande),
                    propose: toPV(action.indexedInformations.montants_versement_propose),
                    accorde: toPV(action.indexedInformations.montants_versement_accorde),
                    attribue: toPV(action.indexedInformations.montants_versement_attribue),
                    realise: toPV(action.indexedInformations.montants_versement_realise),
                    compensation: toPV(action.indexedInformations.montants_versement_compensation),
                },
            }));
        }

        return data;
    }

    static toCommon(entity: OsirisRequestEntity): CommonApplicationDto {
        return {
            dispositif: entity.providerInformations.dispositif,
            exercice: entity.providerInformations.exercise,
            montant_accorde: entity.providerInformations.montantsAccorde,
            montant_demande: entity.providerInformations.montantsDemande,
            objet: (entity.actions || []).map(action => action.indexedInformations.intitule).join(" – ") || "",
            service_instructeur: entity.providerInformations.service_instructeur,
            siret: entity.legalInformations.siret,
            statut: OsirisRequestAdapter.toStatus(entity.providerInformations.status),
        };
    }

    // find if identifier is a disguised Ridet or a native Siret
    static getAssoIdType(identifier: string) {
        // disguised ridet starts with 9900 or 99000
        if (identifier.startsWith("9900")) return Ridet.getName();
        else return Siret.getName();
    }

    // transform disguised Ridet into a valid Ridet
    static cleanRidet(osirisRidet: string) {
        const ridet = osirisRidet.replace(/^99/, "").replace(/^0+/, ""); // ridet is 9 or 10 digits. It removes the starting 9900 or 99000 used by osiris to convert ridet into siret
        if (!Ridet.isRidet(ridet)) {
            throw new Error("Cleaned Ridet is not valid");
        }
        return ridet;
    }

    static getPluriannualYears(entity: OsirisRequestEntity): number[] {
        const startYear = (entity.data as unknown & { Dossier: { "Exercice Début": number; "Exercice Fin": number } })
            .Dossier["Exercice Début"];
        const endYear = (entity.data as unknown & { Dossier: { "Exercice Début": number; "Exercice Fin": number } })
            .Dossier["Exercice Fin"];
        const years: number[] = [];
        for (let start = startYear; start <= endYear; start++) {
            years.push(start);
        }
        return years;
    }

    // return all application cofinancers based on all linked actions
    // return empty string if no cofinancers is found
    static getCofinancers(actions: OsirisActionEntity[]) {
        const cofinancersNames = Array.from(
            actions.reduce((acc, action) => {
                action.indexedInformations.cofinanceurs.split(";").forEach(cofinancer => acc.add(cofinancer));
                return acc;
            }, new Set<string>()),
        )
            .filter(str => str) // remove empty set value from the last trailing comma if exists
            .join("|");
        return cofinancersNames;
    }

    static toApplicationFlat(entity: OsirisRequestEntity, actions: OsirisActionEntity[]): ApplicationFlatEntity {
        const provider = this.PROVIDER_NAME.toLowerCase();
        const budgetaryYear = entity.providerInformations.exercise;
        const applicationProviderId = entity.providerInformations.osirisId;
        const applicationId = `${provider}-${applicationProviderId}`;
        const uniqueId = `${applicationId}-${budgetaryYear}`;
        const assoIdType = this.getAssoIdType(entity.legalInformations.siret);
        const assoId =
            assoIdType === Siret.getName()
                ? entity.legalInformations.siret
                : this.cleanRidet(entity.legalInformations.siret);
        // TODO: make a DTO for OsirisRequest. See #3590
        // @ts-expect-error: dto not available
        const depositDate = GenericParser.ExcelDateToJSDate(entity.data["Dossier"]["Date Reception"]);
        const ej = entity.providerInformations.ej;
        const paymentId = `${assoId}-${ej}-${budgetaryYear}`;
        const cofinancersNames = this.getCofinancers(actions);

        return {
            uniqueId,
            applicationId,
            applicationProviderId,
            provider,
            joinKeyId: entity.providerInformations.compteAssoId,
            joinKeyDesc: `N° dossier de l'outil "Le Compte Asso". Il permet de faire un lien entre la requête OSIRIS et le dossier du Compte Asso.`,
            allocatorName: null,
            allocatorIdType: null,
            allocatorId: null,
            managingAuthorityName: null,
            managingAuthorityId: null,
            managingAuthorityIdType: null,
            instructiveDepartmentName: entity.providerInformations.service_instructeur,
            instructiveDepartmentIdType: null,
            instructiveDepartementId: null,
            beneficiaryEstablishmentId: entity.legalInformations.siret,
            beneficiaryEstablishmentIdType: assoId,
            budgetaryYear,
            pluriannual: entity.providerInformations.pluriannualite === "Pluriannuel",
            pluriannualYears: this.getPluriannualYears(entity).join("|"),
            decisionDate: entity.providerInformations.dateCommission,
            conventionDate: null,
            decisionReference: null,
            depositDate,
            requestYear: depositDate.getFullYear(),
            scheme: entity.providerInformations.dispositif,
            subScheme: entity.providerInformations.sous_dispositif,
            statusLabel: this.toStatus(entity.providerInformations.status),
            object: actions.map(action => action.indexedInformations.intitule).join("|"),
            nature: ApplicationNature.MONEY,
            requestedAmount: entity.providerInformations.montantsDemande,
            grantedAmount: entity.providerInformations.montantsAccorde,
            totalAmount: null,
            ej,
            paymentId,
            paymentCondition: null,
            paymentConditionDesc: null,
            paymentPeriodDates: null,
            cofinancersNames: cofinancersNames,
            cofinancingRequested: cofinancersNames.length > 0,
            cofinancersIdType: null,
            confinancersId: null,
            idRAE: null,
            ueNotification: null,
            subventionPercentage: null,
            updateDate: entity.updateDate,
        } as ApplicationFlatEntity;
    }
}
