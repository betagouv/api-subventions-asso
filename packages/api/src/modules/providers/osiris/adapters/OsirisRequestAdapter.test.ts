const mockLabel = "NORMALIZED_LABEL";
const mockToStatus = jest.fn(() => mockLabel);

import OsirisRequestAdapter from "./OsirisRequestAdapter";
import OsirisEntity from "../../../../../tests/modules/providers/osiris/__fixtures__/OsirisEntities";
import Ridet from "../../../../identifierObjects/Ridet";
import DEFAULT_ASSOCIATION, { RIDET_STR } from "../../../../../tests/__fixtures__/association.fixture";
import Siret from "../../../../identifierObjects/Siret";
import OsirisActionEntity from "../entities/OsirisActionEntity";
import OsirisRequestEntity from "../entities/OsirisRequestEntity";
import { GenericParser } from "../../../../shared/GenericParser";
jest.mock("../../../../shared/GenericParser");

jest.mock("../../providers.adapter", () => ({
    toStatusFactory: () => mockToStatus,
    __esModule: true, // this property makes it work
}));

describe("OsirisRequestAdapter", () => {
    describe("toDemandeSubvention()", () => {
        it("should return a DemandeSubvention", () => {
            const actual = OsirisRequestAdapter.toDemandeSubvention(OsirisEntity);
            expect(actual).toMatchSnapshot();
        });

        it("should set versementKey to undefined if no EJ", () => {
            const entity = JSON.parse(JSON.stringify(OsirisEntity));
            entity.providerInformations.ej = undefined;
            const actual = OsirisRequestAdapter.toDemandeSubvention(entity);
            expect(actual).toMatchSnapshot();
        });

        it("generates status translator", () => {
            const PROVIDER_STATUS = "toto";
            OsirisRequestAdapter.toDemandeSubvention({
                ...OsirisEntity,
                // @ts-expect-error: mock
                providerInformations: { status: PROVIDER_STATUS },
            });
            expect(mockToStatus).toBeCalledWith(PROVIDER_STATUS);
        });

        it("uses status translator", () => {
            const PROVIDER_STATUS = "toto";
            const res = OsirisRequestAdapter.toDemandeSubvention({
                ...OsirisEntity,
                // @ts-expect-error: mock
                providerInformations: { status: PROVIDER_STATUS },
            });
            const actual = res?.statut_label?.value;
            expect(actual).toBe(mockLabel);
        });
    });

    describe("rawToApplication", () => {
        // @ts-expect-error: parameter type
        const RAW_APPLICATION: RawApplication = { data: { foo: "bar" } };
        // @ts-expect-error: parameter type
        const APPLICATION: DemandeSubvention = { foo: "bar" };
        let mockToDemandeSubvention: jest.SpyInstance;

        beforeAll(() => {
            mockToDemandeSubvention = jest.spyOn(OsirisRequestAdapter, "toDemandeSubvention");
            mockToDemandeSubvention.mockReturnValue(APPLICATION);
        });

        afterAll(() => {
            mockToDemandeSubvention.mockRestore();
        });

        it("should call toDemandeSubvention", () => {
            OsirisRequestAdapter.rawToApplication(RAW_APPLICATION);
            expect(mockToDemandeSubvention).toHaveBeenCalledWith(APPLICATION);
        });

        it("should return DemandeSubvention", () => {
            const expected = APPLICATION;
            const actual = OsirisRequestAdapter.rawToApplication(RAW_APPLICATION);
            expect(actual).toEqual(expected);
        });
    });

    describe("toCommon", () => {
        it("returns proper result", () => {
            const INPUT = {
                providerInformations: {
                    dispositif: "DISPOSITIF",
                    exercise: 2023,
                    montantsAccorde: 42,
                    montantsDemande: 43,
                    service_instructeur: "INSTRUCTEUR",
                    status: "NATIVE_STATUS",
                },
                legalInformations: { siret: "123456789" },
                actions: [
                    {
                        indexedInformations: { intitule: "ACTION 1" },
                    },
                    {
                        indexedInformations: { intitule: "ACTION 2" },
                    },
                ],
            };
            // @ts-expect-error mock
            const actual = OsirisRequestAdapter.toCommon(INPUT);
            expect(actual).toMatchSnapshot();
        });
    });

    describe("getAssoIdType", () => {
        it.each`
            osirisRidet
            ${"990001234567891"}
            ${"99001234567891"}
        `("returns ridet", ({ osirisRidet }) => {
            const expected = Ridet.getName();
            const actual = OsirisRequestAdapter.getAssoIdType(osirisRidet);
            expect(actual).toEqual(expected);
        });

        it("returns siret", () => {
            const expected = Siret.getName();
            const actual = OsirisRequestAdapter.getAssoIdType(DEFAULT_ASSOCIATION.siret);
            expect(actual).toEqual(expected);
        });
    });

    describe("cleanRidet", () => {
        it.each`
            osirisRidet
            ${"990001234567891"}
            ${"99001234567891"}
        `("returns well formatted ridet", ({ osirisRidet }) => {
            const expected = "1234567891";
            const actual = OsirisRequestAdapter.cleanRidet(osirisRidet);
            expect(actual).toEqual(expected);
        });

        it("throws error if cleaned ridet is not valid", () => {
            expect(() => OsirisRequestAdapter.cleanRidet("99000123456")).toThrow("Cleaned Ridet is not valid");
        });
    });

    describe("getPluriannualYears", () => {
        it("returns years", () => {
            const expected = [2023, 2024, 2025];
            const actual = OsirisRequestAdapter.getPluriannualYears({
                data: {
                    Dossier: { "Exercice Début": 2023, "Exercice Fin": 2025 },
                },
            } as OsirisRequestEntity);
            expect(actual).toEqual(expected);
        });
    });

    describe("getCofinancers", () => {
        it("returns array of unique cofinancers", () => {
            const expected = [
                "Direction départementale de Saône-et-Loire",
                "DRAC",
                "ARS",
                "Bourgogne-Franche-Comté",
                "Saône-et-Loire",
                "Politque de la ville",
                "CAF",
                "Aides privées",
            ];

            const actual = OsirisRequestAdapter.getCofinancers([
                {
                    indexedInformations: {
                        // copied from real data
                        cofinanceurs:
                            "Direction départementale de Saône-et-Loire;DRAC;ARS;Bourgogne-Franche-Comté;Saône-et-Loire;Politque de la ville;CAF;",
                    },
                } as OsirisActionEntity,
                {
                    indexedInformations: {
                        cofinanceurs: "Politque de la ville;CAF;Aides privées;",
                    },
                } as OsirisActionEntity,
            ]);
            expect(actual).toEqual(expected);
        });

        it("returns empty array if no cofinancers", () => {
            const expected = [];

            const actual = OsirisRequestAdapter.getCofinancers([
                {
                    indexedInformations: {
                        cofinanceurs: "",
                    },
                } as OsirisActionEntity,
                {
                    indexedInformations: {
                        cofinanceurs: "",
                    },
                } as OsirisActionEntity,
            ]);
            expect(actual).toEqual(expected);
        });
    });

    describe("toApplicationFlat", () => {
        let mockGetAssoIdType;
        let mockExcelDateToJSDate;
        let mockCleanRidet;
        let mockGetCofinancers;

        const ACTIONS = [
            {
                indexedInformations: {
                    // copied from real data
                    cofinanceurs:
                        "Direction départementale de Saône-et-Loire;DRAC;ARS;Bourgogne-Franche-Comté;Saône-et-Loire;Politque de la ville;CAF;",
                },
            } as OsirisActionEntity,
        ];

        beforeEach(() => {
            mockGetAssoIdType = jest.spyOn(OsirisRequestAdapter, "getAssoIdType").mockReturnValue(Siret.getName());
            mockExcelDateToJSDate = jest
                .spyOn(GenericParser, "ExcelDateToJSDate")
                .mockReturnValue(new Date("2025-08-04"));
            mockCleanRidet = jest.spyOn(OsirisRequestAdapter, "cleanRidet").mockReturnValue(RIDET_STR);
            mockGetCofinancers = jest
                .spyOn(OsirisRequestAdapter, "getCofinancers")
                .mockReturnValue(["CAF", "ARS", "DRAC"]);
        });

        afterAll(() => {
            [mockGetAssoIdType, mockCleanRidet, mockGetCofinancers].map(mock => mock.mockRestore());
        });

        it("gets identifier type", () => {
            OsirisRequestAdapter.toApplicationFlat(OsirisEntity, ACTIONS);
            expect(mockGetAssoIdType).toHaveBeenCalledWith(OsirisEntity.legalInformations.siret);
        });

        it("format excel date", () => {
            OsirisRequestAdapter.toApplicationFlat(OsirisEntity, ACTIONS);
            // @ts-expect-error: data is defined in mock
            expect(mockExcelDateToJSDate).toHaveBeenCalledWith(OsirisEntity.data["Dossier"]["Date Reception"]);
        });

        it("retrieves the real ridet from the one disguised in siret by osiris", () => {
            mockGetAssoIdType.mockReturnValue(Ridet.getName());
            OsirisRequestAdapter.toApplicationFlat(OsirisEntity, ACTIONS);
            expect(mockCleanRidet).toHaveBeenCalledWith(OsirisEntity.legalInformations.siret);
        });

        it("gets cofinancers", () => {
            OsirisRequestAdapter.toApplicationFlat(OsirisEntity, ACTIONS);
            expect(mockGetCofinancers).toHaveBeenCalledWith(ACTIONS);
        });

        it("returns application flat with default date", () => {
            jest.useFakeTimers().setSystemTime(new Date("2025-08-06"));
            const actual = OsirisRequestAdapter.toApplicationFlat(OsirisEntity, ACTIONS);
            expect(actual).toMatchSnapshot();
            jest.useFakeTimers().useRealTimers();
        });

        it("returns application flat with date from object id", () => {
            const actual = OsirisRequestAdapter.toApplicationFlat(OsirisEntity, ACTIONS);
            expect(actual).toMatchSnapshot();
        });

        it("nullify idVersement if EJ is missing", () => {
            const actual = OsirisRequestAdapter.toApplicationFlat(
                { ...OsirisEntity, providerInformations: { ...OsirisEntity.providerInformations, ej: undefined } },
                ACTIONS,
            );
            expect(actual).toMatchSnapshot();
        });
    });
});
