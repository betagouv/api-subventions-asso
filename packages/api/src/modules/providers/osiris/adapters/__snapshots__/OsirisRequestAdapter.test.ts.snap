// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OsirisRequestAdapter toCommon returns proper result 1`] = `
Object {
  "dispositif": "DISPOSITIF",
  "exercice": 2023,
  "montant_accorde": 42,
  "montant_demande": 43,
  "objet": "ACTION 1 â€“ ACTION 2",
  "service_instructeur": "INSTRUCTEUR",
  "siret": "123456789",
  "statut": "NORMALIZED_LABEL",
}
`;

exports[`OsirisRequestAdapter toDemandeSubvention() should return a DemandeSubvention 1`] = `
Object {
  "actions_proposee": Array [],
  "annee_demande": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "number",
    "value": 2022,
  },
  "contact": Object {
    "email": Object {
      "last_update": 2022-01-01T00:00:00.000Z,
      "provider": "OSIRIS",
      "type": "undefined",
      "value": undefined,
    },
    "telephone": undefined,
  },
  "date_commision": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "object",
    "value": 2022-01-01T00:00:00.000Z,
  },
  "dispositif": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "undefined",
    "value": undefined,
  },
  "ej": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "string",
    "value": "EJ00001",
  },
  "montants": Object {
    "accorde": Object {
      "last_update": 2022-01-01T00:00:00.000Z,
      "provider": "OSIRIS",
      "type": "undefined",
      "value": undefined,
    },
    "demande": Object {
      "last_update": 2022-01-01T00:00:00.000Z,
      "provider": "OSIRIS",
      "type": "undefined",
      "value": undefined,
    },
    "propose": Object {
      "last_update": 2022-01-01T00:00:00.000Z,
      "provider": "OSIRIS",
      "type": "undefined",
      "value": undefined,
    },
    "total": Object {
      "last_update": 2022-01-01T00:00:00.000Z,
      "provider": "OSIRIS",
      "type": "undefined",
      "value": undefined,
    },
  },
  "pluriannualite": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "undefined",
    "value": undefined,
  },
  "service_instructeur": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "undefined",
    "value": undefined,
  },
  "siret": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "string",
    "value": "00000000000001",
  },
  "sous_dispositif": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "undefined",
    "value": undefined,
  },
  "status": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "undefined",
    "value": undefined,
  },
  "statut_label": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "string",
    "value": "NORMALIZED_LABEL",
  },
  "territoires": Array [],
  "versement": Object {
    "acompte": Object {
      "last_update": 2022-01-01T00:00:00.000Z,
      "provider": "OSIRIS",
      "type": "undefined",
      "value": undefined,
    },
    "compensation": Object {
      "n-1": Object {
        "last_update": 2022-01-01T00:00:00.000Z,
        "provider": "OSIRIS",
        "type": "undefined",
        "value": undefined,
      },
      "reversement": Object {
        "last_update": 2022-01-01T00:00:00.000Z,
        "provider": "OSIRIS",
        "type": "undefined",
        "value": undefined,
      },
    },
    "realise": Object {
      "last_update": 2022-01-01T00:00:00.000Z,
      "provider": "OSIRIS",
      "type": "undefined",
      "value": undefined,
    },
    "solde": Object {
      "last_update": 2022-01-01T00:00:00.000Z,
      "provider": "OSIRIS",
      "type": "undefined",
      "value": undefined,
    },
  },
  "versementKey": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "string",
    "value": "EJ00001",
  },
}
`;

exports[`OsirisRequestAdapter toDemandeSubvention() should set versementKey to undefined if no EJ 1`] = `
Object {
  "actions_proposee": Array [],
  "annee_demande": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "number",
    "value": 2022,
  },
  "contact": Object {
    "email": Object {
      "last_update": 2022-01-01T00:00:00.000Z,
      "provider": "OSIRIS",
      "type": "undefined",
      "value": undefined,
    },
    "telephone": undefined,
  },
  "date_commision": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "string",
    "value": "2022-01-01T00:00:00.000Z",
  },
  "dispositif": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "undefined",
    "value": undefined,
  },
  "ej": undefined,
  "montants": Object {
    "accorde": Object {
      "last_update": 2022-01-01T00:00:00.000Z,
      "provider": "OSIRIS",
      "type": "undefined",
      "value": undefined,
    },
    "demande": Object {
      "last_update": 2022-01-01T00:00:00.000Z,
      "provider": "OSIRIS",
      "type": "undefined",
      "value": undefined,
    },
    "propose": Object {
      "last_update": 2022-01-01T00:00:00.000Z,
      "provider": "OSIRIS",
      "type": "undefined",
      "value": undefined,
    },
    "total": Object {
      "last_update": 2022-01-01T00:00:00.000Z,
      "provider": "OSIRIS",
      "type": "undefined",
      "value": undefined,
    },
  },
  "pluriannualite": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "undefined",
    "value": undefined,
  },
  "service_instructeur": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "undefined",
    "value": undefined,
  },
  "siret": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "string",
    "value": "00000000000001",
  },
  "sous_dispositif": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "undefined",
    "value": undefined,
  },
  "status": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "undefined",
    "value": undefined,
  },
  "statut_label": Object {
    "last_update": 2022-01-01T00:00:00.000Z,
    "provider": "OSIRIS",
    "type": "string",
    "value": "NORMALIZED_LABEL",
  },
  "territoires": Array [],
  "versement": Object {
    "acompte": Object {
      "last_update": 2022-01-01T00:00:00.000Z,
      "provider": "OSIRIS",
      "type": "undefined",
      "value": undefined,
    },
    "compensation": Object {
      "n-1": Object {
        "last_update": 2022-01-01T00:00:00.000Z,
        "provider": "OSIRIS",
        "type": "undefined",
        "value": undefined,
      },
      "reversement": Object {
        "last_update": 2022-01-01T00:00:00.000Z,
        "provider": "OSIRIS",
        "type": "undefined",
        "value": undefined,
      },
    },
    "realise": Object {
      "last_update": 2022-01-01T00:00:00.000Z,
      "provider": "OSIRIS",
      "type": "undefined",
      "value": undefined,
    },
    "solde": Object {
      "last_update": 2022-01-01T00:00:00.000Z,
      "provider": "OSIRIS",
      "type": "undefined",
      "value": undefined,
    },
  },
  "versementKey": undefined,
}
`;
